### ISSUE TITLE
formatOnSave re-saves several times

### ISSUE THREAD

**Issue by ghaabor** [Microsoft/vscode-go#1037]
So I've started using the Go extension yesterday and noticed a weird issue. When formatOnSave is on, I push Ctrl-S and the file is saved several times and the undo stack gets crowded with characters getting selected and deselected. It always selects the same characters that were before the cursor when I pressed Ctrl-S. I tried it with all three formatters and they all do the same. 

I think there could be some bigger issue behind this because turning off formatOnSave stops this behavior but the characters in the file still get selected but only once (same here, the character before the cursor).

I use VSCode with Ruby as well and I haven't noticed anything similar. I use the latest Code (1.13.1 I think) and the latest extension. I'm on Ubuntu 17.04.

**Comment by ramya-rao-a** [comment]
That sounds like some other extension is interfering with the save action. We have seen this before.

What extensions do you have? Can you try disabling all extensions, then enable just the Go extension and see if this still persists?

![image](https://user-images.githubusercontent.com/16890566/27299306-c41d2fee-54e0-11e7-8421-01ffba13dbf4.png)

**Comment by wader** [comment]
Hello, i think i have the same issue.
macOS 10.12.5 (16F73)
Code 1.13.1 (1.13.1)
vscode-go 0.6.62

Get same issue (two file writes) when all extensions disabled except vscode-go. But I noticed that if i disable formatOnSave and just use "Format document" i still get one file write, could that be a lead or is that expected behaviour?

**Comment by wader** [comment]
Oh just noticed https://github.com/Microsoft/vscode-go/blob/bca4dd5f31f32ac49da79580c07b4000f06287a3/src/goFormat.ts#L69 is the save needed because the formatting tools can only read from file?

**Comment by ramya-rao-a** [comment]
When you say 2 file writes do you mean the file state indicator next to file name goes from dirty - save - dirty - save?

The line you are pointing at would only cause dirty - save - save

**Comment by wader** [comment]
Ok the go formatting tools support reading from stdin but vscode-go is currently not using it.

So could the issue be that when `onDidSaveTextDocument` is fired the file has already been saved and then the formatting tools are executed and then the document is saved again?

**Comment by wader** [comment]
You mean the dot left of the filename? when i make a change the dot appears. Then when i save it flickers fast and disappears 

Im using https://github.com/cortesi/modd to trigger builds and if i run it in debug mode (`--debug`) i see two writes.
```
18:42:35: notify.Write: "/path/to/main.go"
18:42:35: Delta:
Added: []
Deleted: []
Changed: [/path/to/main.go]
18:42:35: prep: go install
18:42:36: notify.Write: "/path/to/main.go"
>> done (580ms)
18:42:37: daemon: app
>> sending signal hangup
18:42:37: Delta:
Added: []
Deleted: []
Changed: [/path/to/main.go]
18:42:37: go install
18:42:37: daemon: app
exited: signal: hangup
```

**Comment by wader** [comment]
My thinking is that if vscode-go currently uses the go formatting tools by pointing them to the source file path (not using stdin or a temp file) then it would require two save to work. First one to save the maybe not properly formatted file to disk and then run the formatting tool on the file and then second one to save the change to make it properly formatted.

**Comment by wader** [comment]
Sorry maybe i should mention that i don't have any issues withe undo stack. Just the issue with two writes.

**Comment by ramya-rao-a** [comment]
I did try passing the file contents via stain to the formatting tool once before, it caused a bug. I can't remember more on that. Am on vacation right now and so don't have access to my mac. Will check this once I get back.

**Comment by wader** [comment]
Ah ok, and also use `onWillSaveTextDocument` instead? i might try to fix it if i get some free time.

But no hurry, enjoy your vacation!

**Comment by wader** [comment]
@ramya-rao-a do you remember what bug you saw when using stdin?

**Comment by ramya-rao-a** [comment]
@wader I don't remember. Will have to look at the history of the goformat.ts file.

About onWillSaveTextDocument..
The catch here is that if formatting takes more than 750ms (am picking a number here, need to look at the code to get the actual number), then the action is cancelled.

I still haven't come up with a good solution for that. There us another issue tracking it. (will link it once am back from vacation and have something better than my phone to work on)

For now, go ahead and remove the save and pass file contents using stain and send a PR

**Comment by wader** [comment]
Ok, thanks! I've sent a PR to use stdin #1069

**Comment by wader** [comment]
@ramya-rao-a This commit it guess https://github.com/Microsoft/vscode-go/commit/068754e142c7d1ef1b5562e72d50a2cff4346562

**Comment by ramya-rao-a** [comment]
Yes, thats the revert commit.

See https://github.com/Microsoft/vscode-go/pull/1069#issuecomment-313889880 for my thoughts on the usage of "onWillSaveTextDocument" and "stdin"

Also the reason for the double save is not due to https://github.com/Microsoft/vscode-go/blob/bca4dd5f31f32ac49da79580c07b4000f06287a3/src/goFormat.ts#L69. That line is only reached when you explicity run the "Format Document" command.

The actual reason for the second save is https://github.com/Microsoft/vscode-go/blob/bca4dd5f31f32ac49da79580c07b4000f06287a3/src/goMain.ts#L308

**Comment by ramya-rao-a** [comment]
Closing in favor of #540 which will fix this issue as well

**Comment by ramya-rao-a** [comment]
The latest update to the Go extension (0.6.70) has the fix for this issue. There won't be any re-saving of files on formatting
